Tests/PostgresNIOTests/New/PSQLRowStreamTests.swift

```diff
@@ -231,6 +240,196 @@ final class PSQLRowStreamTests: XCTestCase {
+                [self.makeColumnDescription(name: "foo", dataType: .text, format: .binary)],
+                dataSource
+            ),
+            eventLoop: self.eventLoop,
+            logger: self.logger
+        )
+        XCTAssertEqual(dataSource.hitDemand, 0)
+        XCTAssertEqual(dataSource.hitCancel, 0)
+
+        stream.receive([
+            [ByteBuffer(string: "0")],
+            [ByteBuffer(string: "1")]
+        ])
+
+        XCTAssertEqual(dataSource.hitDemand, 0, "Before we have a consumer demand is not signaled")
+        stream.receive(completion: .success("SELECT 2"))
+
+        // attach consumer
+        XCTAssertNoThrow(try stream.drain().wait())
+        XCTAssertEqual(dataSource.hitDemand, 0) // TODO: Is this right?
+    }
+
+    func testDrainBeforeStreamHasFinished() {
+        let dataSource = CountingDataSource()
+        let stream = PSQLRowStream(
+            source: .stream(
+                [self.makeColumnDescription(name: "foo", dataType: .text, format: .binary)],
+                dataSource
+            ),
+            eventLoop: self.eventLoop,
+            logger: self.logger
+        )
+        XCTAssertEqual(dataSource.hitDemand, 0)
+        XCTAssertEqual(dataSource.hitCancel, 0)
+
+        stream.receive([
+            [ByteBuffer(string: "0")],
+            [ByteBuffer(string: "1")]
+        ])
+
+        XCTAssertEqual(dataSource.hitDemand, 0, "Before we have a consumer demand is not signaled")
+
+        // attach consumer
+        let future = stream.drain()
+        XCTAssertEqual(dataSource.hitDemand, 1)
+
+        stream.receive([
+            [ByteBuffer(string: "2")],
+            [ByteBuffer(string: "3")]
+        ])
+        XCTAssertEqual(dataSource.hitDemand, 2)
+
+        stream.receive([
+            [ByteBuffer(string: "4")],
+            [ByteBuffer(string: "5")]
+        ])
+        XCTAssertEqual(dataSource.hitDemand, 3)
+
+        stream.receive(completion: .success("SELECT 2"))
+
+        XCTAssertNoThrow(try future.wait())
+    }
+
+    func testDrainBeforeStreamHasFinishedWhenThereIsAlreadyAConsumer() {
+        let dataSource = CountingDataSource()
+        let stream = PSQLRowStream(
+            source: .stream(
+                [self.makeColumnDescription(name: "foo", dataType: .text, format: .binary)],
+                dataSource
+            ),
+            eventLoop: self.eventLoop,
+            logger: self.logger
+        )
+        XCTAssertEqual(dataSource.hitDemand, 0)
+        XCTAssertEqual(dataSource.hitCancel, 0)
+
+        stream.receive([
+            [ByteBuffer(string: "0")],
+            [ByteBuffer(string: "1")]
+        ])
+
+        XCTAssertEqual(dataSource.hitDemand, 0, "Before we have a consumer demand is not signaled")
+
+        // attach consumers
+        let allFuture = stream.all()
+        XCTAssertEqual(dataSource.hitDemand, 1)
+        let drainFuture = stream.drain()
+        XCTAssertEqual(dataSource.hitDemand, 2)
```

Comment:
I reckon this case won't actually happen at all with the current query functions and technically should only happen if users are misusing something like the postgres-seq by escaping it and trying to consume it weirdly or in a Task. It hits the `case .waitingForAll` code in `drain0()`.

Should i remove it or just let it be incase?

Reply:
I'd personally put a different `preconditionFailure` in `case .waitingForAll ` and try to let users know what they did wrong + delete this test 🤔.

Reply:
Or maybe we should keep that code for the sake of completeness in `drain0()` and having it not be like "only supposed to be used before you want to get the metadata"